xml_loader/shift_generation.py

```diff
@@ -143,70 +145,131 @@ def get_employee_lists(root):
         "employee_contracted_hours": employee_contracted_hours,
     }
 
+def get_demand_pairs(demand, day):
 
-def get_durations(root):
+    start_times = [t + 24 * int(day) for t in demand.start]
+    end_times = [t + 24 * int(day) for t in demand.end]
+    demand_pairs = []
 
-    events = get_events(root)
-    durations = {}
+    for i in range(len(start_times)):
+        demand_pairs.append((start_times[i], end_times[i]))
 
-    # Possible durations are between 6.0 hours and 12.0 hours
-    possible_durations = [t / 4 for t in range(6 * 4, (12 * 4 + 1))]
+    return demand_pairs
+
+
+def get_demand_intervals(demand, day):
+
+    demand_pairs = get_demand_pairs(demand, day)
+    related_intervals = []
+
+    if len(demand_pairs) == 1:
+       related_intervals.append([demand_pairs[0][0], demand_pairs[0][1]])
+    else:
+        for index in range(len(demand_pairs)):
+            temp_related_intervals = [demand_pairs[index][0], demand_pairs[index][1]]
+            for pair in demand_pairs[index + 1:]:
+                if temp_related_intervals[-1] == pair[0]:
+                    temp_related_intervals.append(pair[1])
+                    demand_pairs.remove(pair)
+                else:
+                    break
+            related_intervals.append(temp_related_intervals)
+            if index is len(demand_pairs) - 1:
+                break
+    return related_intervals
 
-    for t in events:
-        for dur in possible_durations:
-            if t + dur in events:
-                try:
-                    durations[t].append(dur)
-                except:
-                    durations[t] = [dur]
-    return durations
+
+def get_daily_demand_intervals(root):
+
+    daily_demand = get_days_with_demand(root)
+    daily_demand_intervals = {}
+
+    for day in daily_demand:
+        daily_demand_intervals[day] = get_demand_intervals(daily_demand[day], day)
+
+    return(daily_demand_intervals)
+
+
+def combine_demand_intervals(root):
+
+    daily_demand_intervals = get_daily_demand_intervals(root)
+    interval_list = []
+    combined_demand_intervals = []
+
+    for day in daily_demand_intervals:
```

Comment:
You could possibly simplify this to `for daily_demand in daily_demand_intervals.values()`